pkg/services/ngalert/state/state.go

```diff
@@ -165,6 +165,32 @@ func (c StateTransition) Changed() bool {
 	return c.PreviousState != c.State.State || c.PreviousStateReason != c.State.StateReason
 }
 
+type StateTransitions []StateTransition
+
+// StaleStates returns the subset of StateTransitions that are stale.
+func (c StateTransitions) StaleStates() StateTransitions {
+	var result StateTransitions
+	for _, t := range c {
+		if t.IsStale() {
+			result = append(result, t)
+		}
+	}
+	return result
+}
+
+// UpdateLastSentAt returns the subset StateTransitions that need sending and updates their LastSentAt field.
+// Note: This is not idempotent, running this twice can (and usually will) return different results.
+func (c StateTransitions) UpdateLastSentAt(evaluatedAt time.Time, resendDelay, resolvedRetention time.Duration) StateTransitions {
```

Comment:
I do not think I really like it because StateTransitions is a public model and exposing methods that create side-effects is not a good idea. 
What do you think about making it a function somewhere close to ProcessEvalResult where it's called? Or at very least make it package private.

Reply:
Agreed, this is a holdover from a previous revision where it was still called from `alert_rule`. I can either inline it or add it as a private method on state mananger (since it uses `st.ResendDelay` and `st.ResolvedRetention`).

For now, I've [updated](https://github.com/grafana/grafana/pull/89427/commits/b0957b047d078f4cfdbc62fa85fc92aa026c1a4f) it to be a state manager method, but don't have strong feelings one way or the other.